[LLDB] Add AST node classes, functions, etc. for Data Inspection Lang…

cmtice opened the pull request:
…uage (DIL).

The Data Inspection Language (DIL), described in
https://discourse.llvm.org/t/rfc-data-inspection-language/69893 includes its own parser and expression evaluator.

This change defines the AST nodes, classes and functions which are used by the DIL parser and expression evaluator. It also adds the .ebnf file documenting the (current) expression language handled by the DIL parser and expression evaluator.

cmtice committed:
[LLDB] Add AST node classes, functions, etc. for Data Inspection Language (DIL).

The Data Inspection Language (DIL), described in
https://discourse.llvm.org/t/rfc-data-inspection-language/69893 includes
its own parser and expression evaluator.

This change defines the AST nodes, classes and functions which are used by the
DIL parser and expression evaluator. It also adds the .ebnf file documenting
the (current) expression language handled by the DIL parser and expression
evaluator.

adrian-prantl commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+(* LLDB Debug Expressions, a subset of C++ *)
+(* Insired by https://www.nongnu.org/hcb *)
+
+expression = assignment_expression ;
```
Nice to actually have the grammar documented!

werat commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+                   | postfix_expression "->" id_expression
+                   | postfix_expression "++"
+                   | postfix_expression "--"
+                   | static_cast "<" type_id ">" "(" expression ")" ;
+                   | dynamic_cast "<" type_id ">" "(" expression ")" ;
+                   | reinterpret_cast "<" type_id ">" "(" expression ")" ;
```
```suggestion
                   | static_cast "<" type_id ">" "(" expression ")"
                   | dynamic_cast "<" type_id ">" "(" expression ")"
                   | reinterpret_cast "<" type_id ">" "(" expression ")" ;
```

cmtice replied:
Done.

werat commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+nested_name_specifier = type_name "::"
+                      | namespace_name '::'
+                      | nested_name_specifier identifier "::"
+                      | nested_name_specifier simple_template_id "::"
```
```suggestion
                      | nested_name_specifier simple_template_id "::" ;
```

cmtice replied:
Done.

cmtice replied:
Done.

werat commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+builtin_func_name = "__log2" ;
+
+builtin_func_argument_list = builtin_func_argument
+                           | builtin_func_argument_list "," builtin_func_argument
```
```suggestion
                           | builtin_func_argument_list "," builtin_func_argument ;
```

werat commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+                   | postfix_expression "->" id_expression
+                   | postfix_expression "++"
+                   | postfix_expression "--"
+                   | static_cast "<" type_id ">" "(" expression ")" ;
+                   | dynamic_cast "<" type_id ">" "(" expression ")" ;
+                   | reinterpret_cast "<" type_id ">" "(" expression ")" ;
```
Not sure how I made those mistakes over time, but the semicolons are wrong in some places 😅

cmtice committed:
[LLDB] Add AST node classes, functions, etc. for Data Inspection Language

Fix semicolons in the .ebnf file.

Michael137 commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+cast_expression = unary_expression
+                | "(" type_id ")" cast_expression ;
+
+unary_expression = postfix_expression
```
Do we really want to allow state-changing expressions in the `frame var` language?

werat replied:
It is pretty useful to be able to easily change the program state and afaik it's already possible with `expr` command. Generally other debugger allow this as well (e.g. Visual Studio), so I think it's a reasonable thing to do. However, for increments specifically there's a deeper reason here (and I would agree that they probably shouldn't be available for general use). See my big comment below for justification.

labath replied:
I would love to have a convenient way to change a variable value without invoking the full expression parser (or going into python), but I think there's a discussion to be made about how to expose that functionality. For example, the current envisioned entry points into this code (the "frame variable" CLI cmd, and `SBFrame.GetValueForVariablePath` don't exactly sound like they ought to be modifying the variable in the process of "getting" it).

cmtice replied:
On the other hand, many users confuse or conflate "frame variable" with "p" (the short-cut for the expression evaluator), the "p" certainly allows updating the variable value...

Michael137 reviewed (commented):
Haven't gone through this in detail yet (having an example of how these structures will be used would be helpful), but something I was curious about is whether this `DIL` is going to be used in the implementation of a new command, or if we're going to extend the `frame var` language? I didn't see a concrete conclusion on this in the RFC

Michael137 commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+
+pointer_literal = "nullptr" ;
+
+builtin_func = builtin_func_name "(" [builtin_func_argument_list] ")" ;
```
What are builtins here? A set of functions that LLDB knows how to execute? How are those chosen?

werat replied:
`lldb-eval` (DIL's predecessor) supports exactly two builtin functions at the moment: `__log2` and cryptic `__findnonnull` (https://github.com/google/lldb-eval/blob/323d48f0b06d7dcf6ed70cdde17b42285450a32a/lldb-eval/parser.cc#L619). The latter is commonly used in NatVis for complex visualisers, e.g. https://github.com/EQEmu/Server/blob/7918fed81cae8700d73319d4bed5c9f265dbbf3d/libs/perlbind/src/perlbind.natvis#L88

Michael137 commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+
+typedef_name = identifier ;
+
+simple_template_id = template_name "<" [template_argument_list] ">" ;
```
Curious what the use-case for instantiating templates is in the DIL. I guess referencing static data members of different instantiations?

Michael137 replied:
Also how will this be implemented? I guess the simple-template-names infrastructure could help here?
AFAIU in the expression evaluator templates would ideally be supported by either using modules or extending DWARF in a way that allows us to construct more faithful ASTs.

werat replied:
Typically this is used in casts like `((Foo<int>)some_ptr)->field`. In the world of dynamic typing this is not necessary, however VisualStudio debugger doesn't support dynamic typing like this, so everyone is casting things.

Michael137 commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,165 @@
+                   | postfix_expression "--"
+                   | static_cast "<" type_id ">" "(" expression ")"
+                   | dynamic_cast "<" type_id ">" "(" expression ")"
+                   | reinterpret_cast "<" type_id ">" "(" expression ")" ;
```
Wouldn't `C-style` casts suffice?

werat replied:
See the reply above, C++-style casts are often used in NatVis (visual studio debugger), which `lldb-eval` originally aimed to support.

cmtice commented:
> Haven't gone through this in detail yet (having an example of how these structures will be used would be helpful), but something I was curious about is whether this `DIL` is going to be used in the implementation of a new command, or if we're going to extend the `frame var` language? I didn't see a concrete conclusion on this in the RFC

The plan is for the DIL implementation to eventually replace the current 'frame variable' command (see Milestone 1 in the implementation plan).  Once that is done,  it will be expanded to also replace calls to GetValueForExpressionPath, for parsing & handling pieces of LLDB's data formatters.

I couldn't put the whole implementation into a single PR, as it would have been much too big to review (even some of the pieces are going to be pretty big). So I broke it up into: the helper functions (which have already been committed now); the AST (this PR); and then three more PR's, each one coming after the preceding one is approved: the DIL Parser (a recursive-descent parser), which constructs (and returns) the appropriate DIL AST; the DIL Evaluator, which takes a DIL AST, evaluates  it in the current context (frame), and returns an appropriate ValueObjectSP;  and the code to actually hook this up to the 'frame variable' command.  I have the whole thing implemented (complete with test cases). If you really want to dig through the complete code, you can see it at https://github.com/cmtice/llvm-project/tree/DIL-work-new/ (note that I will be cleaning up the Parser & Evaluator code before actually being put into a PR).  The current implementation there still has both implementations for 'frame variable' (original and DIL), and with 'original' being the default and 'DIL' being turned on by an option.

labath commented:
I've been trying to stay out of this, as I don't think I can be objective here, but since I've been summoned, here's what I think. :)

For me, the worst possible outcome of this is an incomplete transition -- having a third (maybe fourth, since I sort of also count dwim-print) expression-like syntax, which is "almost" complete, but for whatever reason (technical or political) can't replace the previous one. With that in mind, I'd try to stage this in a way that minimizes this risk. Building the whole thing in layers from the bottom up is not the best way to achieve that, I think.

If I were doing this, I'd prioritise replacing the current "frame var" implementation as soon as possible. Ideally, the steps would be something like:
1. implement the minimal necessary functionality to match current "frame var" feature set. (I think that's `&`, `[]`, `*` and `->`)
2. flip the switch
3. remove the old implementation
4. add other features, one group of related operators at a time

Besides managing that risk (if this fades out for whatever reason, there's less dead code around), I think this would also make it easier to review, as we could have patches like "add operator/ to the DIL", which would focus solely on that, and also come with associated tests. I think it would also make it easier to discuss things like whether we want to add modifying operations (operator++) or obviously language-specific features (dynamic_cast), both of which I think are very good questions.

walter-erquinigo commented:
I second everything that Pavel says. I think this would be the best approach. The existing frame var implementation is not that "smart", so it should be practical to fully replace it with the new DIL in its first stage. 
An additional consideration is that lldb-dap, which is being more and more frequently used, relies on `frame var` for executing most expressions. `expr` is just a fallback. Therefore, replacing `frame var` with DIL right away has the benefit of getting tested right away by all the lldb-dap users.

Michael137 commented:
> I have the whole thing implemented (complete with test cases). If you really want to dig through the complete code, you can see it at https://github.com/cmtice/llvm-project/tree/DIL-work-new/ (note that I will be cleaning up the Parser & Evaluator code before actually being put into a PR).

Thanks! that'll be a useful reference

> Besides managing that risk (if this fades out for whatever reason, there's less dead code around), I think this would also make it easier to review, as we could have patches like "add operator/ to the DIL", which would focus solely on that, and also come with associated tests. I think it would also make it easier to discuss things like whether we want to add modifying operations (operator++) or obviously language-specific features (dynamic_cast), both of which I think are very good questions.

Makes sense!

werat commented:
Providing more context for answers to Michael's questions.

The original motivation for `lldb-eval` (https://github.com/google/lldb-eval, which is a predecessor for DIL) was to support the evaluation of expressions used in NatVis rules. [NatVis](https://learn.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2022) visualizers support complex logic, including C++-like casts, restricted function calls and state-changing operations like increment/decrement.

To be able to evaluate those expressions as is, lldb-eval had to support all of the commonly used features, hence the extensive and somewhat selective list of supported features :)

For example, `reinterpret_cast<>` is commonly used to cast pointers to structs/classes in order to access their fields (example -- https://github.com/godotengine/godot/blob/cae2f853dcd1ecc26ca68de08cec62089dee1f26/platform/windows/godot.natvis#L146). Visual Studio NatVis evaluator doesn't support dynamic typing (e.g. resolving `foo->bar`, if `bar` is defined in a subclass), therefore the casts are necessary. In the world of dynamic typing this may be not necessary.

State altering operations like increment/decrement are used in advanced visualizers like `<CustomListItems>` for control flow -- e.g. you can define a loop variable, then increment it and break out of the loop under some condition. A real life example -- https://github.com/GaijinEntertainment/DagorEngine/blob/8b340d10814f06aa364e75769d406ec7373c5574/prog/_jBuild/_natvis/dagor_types.natvis#L746. It's not visible in the grammar, but `lldb-eval` has additional knobs for controlling when state changing is allowed. If I recall correctly, this can be done in two cases:
1) the variable being incremented is a "context variable", e.g. an artificial SBValue passed to the expression as context, which doesn't exist in the program. See some explanation and example here -- https://werat.dev/blog/blazing-fast-expression-evaluation-for-c-in-lldb/#maintaning-state
2) when side effects are explecitely allowed via options -- https://github.com/google/lldb-eval/blob/323d48f0b06d7dcf6ed70cdde17b42285450a32a/lldb-eval/api.h#L77. In our debugger this was disabled for all visualizer-related evaluations and enabled for one-off interactive evaluations, basically mimicking the behaviour of Visual Studio debugger.

adrian-prantl commented:
A high-level concern that I have with this is the grammar is full of C++-specific nodes, which is not something I would like to see in living in `Core`. Have we previously reached any consensus on how we want to handle different languages in DIL?
I would be more comfortable with moving most of this into the `Plugins/Language/CPlusPlus` directory and having only a generic, language-agnostic driver live in `Core`.

jimingham commented:
> A high-level concern that I have with this is the grammar is full of C++-specific nodes, which is not something I would like to see in living in `Core`. Have we previously reached any consensus on how we want to handle different languages in DIL? I would be more comfortable with moving most of this into the `Plugins/Language/CPlusPlus` directory and having only a generic, language-agnostic driver live in `Core`.

My .02:

The problem here is that the commands that use the DIL often have no idea what language the path expression they are being asked to parse is going to end up being.  

`(lldb) target var some_DIL_expression
`

could find C/ObjC/C++/Rust/Swift variables that match that expression.

The idea for the DIL is to have enough syntax to express sensible data lookup and presentation that would cover the variety of languages we support.  That's going to end up looking like the languages we are using it to model, but it shouldn't be driven by them.  I think that's better than having the DIL have language specific variants and the user somehow has to know to pick the right one for the context.

cmtice committed:
[LLDB] Add AST node classes, functions, etc. for Data Inspection Language

Remove all the pieces for adding functionality beyond what the current
default "frame variable" implementation does. Update the .ebnf file to
reflect the current simplified version of DIL.

cmtice commented:
As requested, I have updated this PR to only include the pieces needed to reproduce current (default) 'frame variable' behavior with DIL.  I have also updated the BNF file accordingly. 

I agree with Jim re the DIL language: We should only have a single language definition, and it can be a superset of the languages  it supports. So there may be parts of it that belong to particular languages, but that does not mean it supports those languages exclusively.

cmtice committed:
Fix clang format issue.

cmtice commented:
BTW, I have verified that this stripped down version passes all the frame variable tests in LLDB.

labath reviewed (commented):
> BTW, I have verified that this stripped down version passes all the frame variable tests in LLDB.

That's cool. Just to confirm, have you looked at replacing `target variable` as well? It uses the same language as "frame var" under the hood, which  but it has a somewhat different starting point (it basically ignores the local scope and looks only at globals), which means it may need some special handling in the new parser.

> I agree with Jim re the DIL language: We should only have a single language definition, and it can be a superset of the languages it supports. So there may be parts of it that belong to particular languages, but that does not mean it supports those languages exclusively.

This direction makes sense to me, but I think that's all the more reason to be conservative/cautious about adding new features to the language. We can't just put every possible feature of every language into it, as we'd end up with a unmaintainable mess.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  explicit operator bool() const { return type.IsValid(); }
+};
+
+/// This determines if the type is a shared, unique or weak pointer, either
+/// from stdlibc++ or libc+++.
+bool IsSmartPtrType(CompilerType type);
+
+/// Finds the member field with the given name and type, stores the child index
+/// corresponding to the field in the idx vector and returns a DILMemberInfo
+/// struct with appropriate information about the field.
+DILMemberInfo GetFieldWithNameIndexPath(lldb::ValueObjectSP lhs_val_sp,
+                                        CompilerType type,
+                                        const std::string &name,
+                                        std::vector<uint32_t> *idx,
+                                        CompilerType empty_type,
+                                        bool use_synthetic, bool is_dynamic);
+
+std::tuple<DILMemberInfo, std::vector<uint32_t>>
+GetMemberInfo(lldb::ValueObjectSP lhs_val_sp, CompilerType type,
+              const std::string &name, bool use_synthetic);
```
They are all implemented in the parser file (which I did not include in this PR).  IsSmartPtr is called in both the DIL parser and the DIL evaluator; GetMemberInfo is called in both the DIL parser and the DIL AST implementation.  Since the DILAST.h file needs to be included in both the parser and the evaluator (and the AST implementation), I put function declarations for functions called in multiple files into DILAST.h.  I can remove them if you want, but will then need to add them back in when for the parser PR.  Let me know what you would prefer.

Initially GetFieldWithNameIndexPath was also called from multiple files, but it looks like it's now only called from the parser, so I will move that function declaration to the parser header file.

cmtice replied:
Moved the declarations to the header files for the parts the define them (not the AST file).

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  kSmartPtrToPtrDecay
+};
+
+/// The C-Style casts allowed by DIL.
+enum class CStyleCastKind {
+  kArithmetic,
+  kEnumeration,
+  kPointer,
+  kNullptr,
+  kReference,
+};
```
Frame var doesn't allow users to specify casts of variables, but it does do some implicit type conversions/casting when evaluating variables, e.g. arrays-to-pointers, and various implicit type promotions & conversions. The DIL stripped down parser/evaluator still needs to do these promotions & conversions, and uses these for that work.

cmtice replied:
Renamed this to TypePromotionCastKind; removed the non-type-promotion cast values.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+/// name/type pairs.
+bool IsContextVar(const std::string &name);
+
+/// Checks to see if the CompilerType is a Smart Pointer (shared, unique, weak)
+/// or not. Only applicable for C++, which is why this is here and not part of
+/// the CompilerType class.
```
I don't understand what you're suggesting here. How would using the synthetic dereference operation help to determine if a pointer variable is a 'smart' pointer vs a regular one?

(I'll remove the second declaration; as mentioned above, this function is implemented in the parser file).

labath replied:
It wouldn't. I assumed that this was used to determine support dereferencing of smart pointers. If that's not true, then my question is: Why do we need to know if something is a (c++) smart pointer?

cmtice replied:
When we go to dereference a pointer we need to know whether or not it's a smart pointer, because if it's a smart pointer we need to get at the pointer inside before we can dereference it. So we use IsSmartPtrType to determine whether or not to insert this layer of unwrapping (converting the smart ptr into a normal ptr).

cmtice replied:
Still working on how to handle smart pointers without special treatment or testing for them; but in the meantime, moved this function declaration to the parser code (where the function definition is, and where it's called), so it's no longer part of this PR.

cmtice commented:
> > BTW, I have verified that this stripped down version passes all the frame variable tests in LLDB.
> 
> That's cool. Just to confirm, have you looked at replacing `target variable` as well? It uses the same language as "frame var" under the hood, which but it has a somewhat different starting point (it basically ignores the local scope and looks only at globals), which means it may need some special handling in the new parser.
> 

No, I have not looked at 'target variable' at this time.

> > I agree with Jim re the DIL language: We should only have a single language definition, and it can be a superset of the languages it supports. So there may be parts of it that belong to particular languages, but that does not mean it supports those languages exclusively.
> 
> This direction makes sense to me, but I think that's all the more reason to be conservative/cautious about adding new features to the language. We can't just put every possible feature of every language into it, as we'd end up with a unmaintainable mess.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+struct DILMemberInfo {
+  std::optional<std::string> name;
+  CompilerType type;
+  bool is_bitfield;
+  uint32_t bitfield_size_in_bits;
```
Some bitfields have an undefined size, e.g. result of a ternary operation. So I'd rather keep these two fields separate.

cmtice replied:
Made the bitfield_size_in_bits field optional, and removed is_bitfield.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  bool is_dynamic;
+  lldb::ValueObjectSP val_obj_sp;
+
+  explicit operator bool() const { return type.IsValid(); }
```
Could you please elaborate? I don't quite understand what you are suggesting here.

cmtice replied:
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+  return type.IsReferenceType() ? type.GetNonReferenceType() : type;
+}
+
+static std::unordered_map<std::string, CompilerType> context_args;
+
+bool IsContextVar(const std::string &name) {
+  return context_args.find(name) != context_args.end();
+}
```
When you say it needs to be scoped, you mean put inside a dil namespace?

This is used by the stripped down parser when it's handling identifiers.

cmtice replied:
On further examination realized that this was how the original code (lldb-eval) was storing/referencing LLDB's convenience variables.  Have removed this, as it's not needed for the DIL code, which can access the convenience variables in other ways.

cmtice commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,43 @@
+
+numeric_literal = ? clang::tok::numeric_constant ? ;
+
+boolean_literal = "true" | "false" ;
+
+pointer_literal = "nullptr" ;
```
Would it be acceptable for me to replace the quoted strings with something like: true_token, false_token, nullptr_token?

cmtice replied:
Removed the quoted strings from the grammar for now.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+
+private:
+  CompilerType m_type;
+  std::variant<llvm::APInt, llvm::APFloat, bool, std::vector<char>> m_value;
```
The Scalar class looks like its designed to handle only ints & floats.  This (LiteralNode) ALSO supports bools and string literals.  I might be able to extend Scalar to handle bools, but I'm not sure it makes sense to expand it to handle string literals, too, in which case I'm not sure what you would want me to do here?

cmtice replied:
Had to split the LiteralNode class into two classes: ScalarLiteralNode and StringLiteralNode. (Done).

cmtice commented:
> On Jul 10, 2024, at 9:22 PM, cmtice ***@***.***> wrote: @cmtice commented on this pull request. In lldb/include/lldb/Core/DILAST.h <[#95738 (comment)](https://github.com/llvm/llvm-project/pull/95738#discussion_r1673380259)>: > +/// Checks to see if the CompilerType is a Smart Pointer (shared, unique, weak) +/// or not. Only applicable for C++, which is why this is here and not part of +/// the CompilerType class. When we go to dereference a pointer we need to know whether or not it's a smart pointer, because if it's a smart pointer we need to get at the pointer inside before we can dereference it. So we use IsSmartPtrType to determine whether or not to insert this layer of unwrapping (converting the smart ptr into a normal ptr).
> I'm not sure it's the job of the DIL to do this sort of unwrapping. Then it would have to support all the ways this might need to be done in every language we might end up supporting. This seems more a job for the data formatters provided by the various languages. That's why Pavel was asking about the special deference Synthetic Child Provider API, I think. That provides a way for various languages to encode this sort of behavior w/o it having to be know in advance by lldb. Jim
> […](#)
> — Reply to this email directly, view it on GitHub <[#95738 (comment)](https://github.com/llvm/llvm-project/pull/95738#discussion_r1673380259)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW4J5FKAPSUTJYXSJODZLYB7TAVCNFSM6AAAAABJNJ3CAKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNZQHA4DAOJYGE>. You are receiving this because you are on a team that was mentioned.

I can use the machinery that's already there in LLDB for dereferencig the smart pointers. But I can't treat smart pointers exactly the same as normal pointers -- among other things a ValueObject whose CompilerType is a smart pointer returns False for IsPointerType(), which causes errors in the code for handling normal pointers.  So as far as I can tell I still need my function (declared in DILAST.h) for telling me whether or not a type is a smart pointer type (which is what Pavel's original comment was about).

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-ast

cmtice committed:
[LLDB] Add AST node classes & functions for Data Inspection Language

Address review comments:
Remove decls for functions not defined or used in AST files; add
DIL namespace, code cleanups, etc.

cmtice commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -0,0 +1,43 @@
+                   | boolean_literal
+                   | pointer_literal
+                   | id_expression
+                   | "this" ;
```
Removed the quoted strings from the grammar for now.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_DIL_AST_H_
```
Done

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+#include "llvm/ADT/APFloat.h"
+#include "llvm/ADT/APInt.h"
+
+namespace lldb_private {
```
Done (namespace "DIL")

cmtice replied:
Done.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  kMemberOfNode,
+  kArraySubscriptNode,
+  kUnaryOpNode,
+  kSmartPtrToPtrDecay
```
Done (changed the 'k' prefixes to 'e' instead).

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+class IdentifierInfo {
+private:
+  using MemberPath = std::vector<uint32_t>;
+  using IdentifierInfoPtr = std::unique_ptr<IdentifierInfo>;
```
Done.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+class DILASTNode {
+public:
+  DILASTNode(clang::SourceLocation location) : location_(location) {}
+  virtual ~DILASTNode() {}
```
Done.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  virtual uint32_t bitfield_size() const { return 0; }
+  virtual CompilerType result_type() const = 0;
+
+  virtual DILNodeKind what_am_i() const = 0;
```
I've added RTTI stuff & replaced what_am_i calls with llvm::Isa<...>. 
I can't remove most of the is_* functions because they can cut across DILAST node types. Removed the ones I could.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+
+  virtual DILNodeKind what_am_i() const = 0;
+
+  clang::SourceLocation location() const { return location_; }
```
Done.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  // The expression result type, but dereferenced in case it's a reference. This
+  // is for convenience, since for the purposes of the semantic analysis only
+  // the dereferenced type matters.
+  CompilerType result_type_deref() const;
```
Done.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  CompilerType result_type_deref() const;
+
+private:
+  clang::SourceLocation location_;
```
Done.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,446 @@
+  clang::SourceLocation location_;
+};
+
+using ParseResult = std::unique_ptr<DILASTNode>;
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+  if (!target)
+    return lldb::ValueObjectSP();
+
+  lock = std::unique_lock<std::recursive_mutex>(target->GetAPIMutex());
```
Decided locking wasn't really needed. Removed locking from this function and renamed it to GetDynamicOrSyntheticValue.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+LookupStaticIdentifier(lldb::TargetSP target_sp,
+                       const llvm::StringRef &name_ref,
+                       ConstString unqualified_name) {
+  // List global variable with the same "basename". There can be many matches
+  // from other scopes (namespaces, classes), so we do additional filtering
+  // later.
```
I've tried to do what you asked? Not sure if I succeeded.

labath replied:
It's definitely better, but I'd still like to understand what is the purpose of the subsequent filtering (lines 229--235). If we didn't have that, we could just pass 1 as the "limit" to FindGlobalVariables and potentially speed things up a lot.

(Somewhat tangentially, I'm also wondering if this function belongs in the "AST" part of DIL. I think I could see a case for putting it into the parser.)

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+      }
+    }
+
+    if (result_type_list.size() == 0) {
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+
+    // Full match is always correct if we're currently in the global scope.
+    if (full_match.IsValid())
+      return full_match;
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+  }
+
+  CompilerType empty_type;
+  return empty_type;
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+    // Check for global vars, which might start with '::'.
+    if (str_ref_name.size() > 2 && str_ref_name[0] == ':' &&
+        str_ref_name[1] == ':')
+      str_ref_name = str_ref_name.drop_front(2);
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+      str_ref_name = str_ref_name.drop_front(2);
+
+    ConstString tmp_name(str_ref_name);
+    if (tmp_name == name)
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+    if (var_sp->GetName() == name) {
+      exact_match = var_sp;
+      break;
+    }
```
Done.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,468 @@
+    Target *target = ctx_scope->CalculateTarget().get();
+    Process *process = ctx_scope->CalculateProcess().get();
+    if (target && process) {
+      Process::StopLocker stop_locker;
```
Turns out it's not actually needed here. Have removed it.

labath commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,560 @@
+         val->GetVariable()->NameMatches(ConstString(name_ref))))
+      return val;
+  }
+  lldb::ValueObjectSP empty_obj_sp;
+  return empty_obj_sp;
```
`return nullptr` or `return {}` works just fine.

labath commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,560 @@
+CompilerType
+ResolveTypeByName(const std::string &name,
+                  std::shared_ptr<ExecutionContextScope> ctx_scope) {
+  // Internally types don't have global scope qualifier in their names and
+  // LLDB doesn't support queries with it too.
+  llvm::StringRef name_ref(name);
+  bool global_scope = false;
+
+  if (name_ref.starts_with("::")) {
+    name_ref = name_ref.drop_front(2);
+    global_scope = true;
+  }
```
I'm not sure that's true, or at least "true anymore". I'm pretty sure I can look up types with leading "::" now.

cmtice replied:
Sometimes it works and sometimes it doesn't. It appears that I can prefix plain global vars with '::', but I can't (e.g.) prefix namespaces with '::'.  So current frame var works with '::globalVar', and also on 'ns::globalVar', but fails on '::ns::globalVar'.

I am unclear if there is any work or change you are requesting here, other than commenting on the comment?

labath commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,560 @@
+    }
+  }
+
+  // We've found multiple types, try finding the "correct" one.
+  CompilerType full_match;
+  std::vector<CompilerType> partial_matches;
+
+  for (uint32_t i = 0; i < result_type_list.size(); ++i) {
+    CompilerType type = result_type_list[i];
+    llvm::StringRef type_name_ref = type.GetTypeName().GetStringRef();
+    ;
+
+    if (type_name_ref == name_ref)
+      full_match = type;
+    else if (type_name_ref.ends_with(name_ref))
+      partial_matches.push_back(type);
+  }
+
+  // Full match is always correct.
+  if (full_match.IsValid())
+    return full_match;
+
+  if (!global_scope) {
+    // We're looking for type, but there may be multiple candidates and which
+    // one is correct may depend on the currect scope. For now just pick the
+    // most "probable" type (pick a random one). TODO: Try to find a better way
+    // to do this.
+    if (partial_matches.size() > 0)
+      return partial_matches.back();
+  }
```
And if that's true, then maybe this filtering isn't needed?

labath reviewed (commented):
I still have a lot of comments, but I think we're making progress.

The most important questions are about the parts where we appear to be reimplementing some existing lldb functionality. Basically all of the functionality for looking up names (for members, variables, types, etc.) appears to be more complex than necessary. For each of those things, we already have lldb functions, which one would (perhaps naively) expect to just call and have them do the right thing. I'd like to understand why that isn't the case, and whether we can do something about it (there are certainly bugs in those functions, and if that's the case it'd be better to fix those bugs than to work around them here).

At this point, perhaps the only way to figure that out is to strip them and see what breaks.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,459 @@
+    eThisKeyword,
+  };
+
+  static IdentifierInfoUP FromValue(lldb::ValueObjectSP value_sp) {
```
Just to make sure I'm not misunderstanding you: You would like me to change the parameter for the FromValue function to be a ValueObject& rather than a ValueObjectSP?

labath replied:
Yes, that was they idea, particularly if the callers already know they are dealing with a valid object. I want to avoid needing to check the pointer validity at every level in the stack.

cmtice replied:
I went back and looked at this a little closer; when the user requests a nonexistent variable name (makes a mistake), this function can be passed an empty ValueObjectSP; the parser checks to see if the IdentifierInfo object that gets returned passes the IsValid check, and that's when the parser either continues parsing or tells the user they've made an invalid request. So no, I can't really change the parameter here on the assumption that the function always gets a valid object.

labath replied:
Hmm... this ties in to the IsValid discussion we had on a different object (DILMemberInfo). Since all of the factory functions creating these objects return a pointer, I think that having an extra IsValid() method is unnecessary and/or confusing as the (in)validity of the result could be indicated by the (in)validity of the pointer.
That I think would also make it natural to move this check to the caller, as it could do something like:
```
if (value_sp) return IdentifierInfo::FromValue(value_sp);
return nullptr;
```

jimingham replied:
I didn't look further into the code to see if you actually treat ValueObjects this way, but just to make sure this is clear, IsValid is a necessary but not sufficient check to see whether you can use an ValueObject.  You also have to check GetError().Success().  If you think about the case of a variable request for a variable that is defined in the block containing the PC, but is optimized out at the current PC.  That is a valid SBValue, since it carries a useful error and might have a value after you step again.  But it doesn't have a value you can use at this stop.

cmtice replied:
Ack.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,459 @@
+private:
+  CompilerType m_type;
+  Scalar m_value;
+  bool m_is_literal_zero;
```
I don't think so? I don't think having a special node type for a literal zero would simplify very much, and it would mean more code duplication (for handling the new node type). But I could be wrong...

labath replied:
Hmm... actually, lets take a step back. Is it ever possible to have a `ScalarLiteralNode`, whose value is zero, but it is not a **literal** zero?

cmtice replied:
I think so: If this were representing a Boolean false.

labath replied:
Okay, so maybe the implementation of "literal zero" can be `m_scalar.IsZero() && !m_type.IsBoolean()`

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,459 @@
+private:
+  CompilerType m_type;
+  std::vector<char> m_value;
+  bool m_is_literal_zero;
```
Whoops; it can't. This is a copy/past error (sorry!).

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,459 @@
+
+private:
+  CompilerType m_type;
+  std::vector<char> m_value;
```
It sort of makes more sense for a string literal to be an array (or vector) of characters.  But I could make this a std::string if you really want me to?

labath replied:
A string *is* a vector of characters, but with extra operations that make sense for operating on strings. If I see vector<char>, I start to think whether there's something subtle going on, like maybe this representing a raw memory dump of something. While it's not possible to make a direct comparison to clang's StringLiteral class (because it uses tail allocations to store the string), its interface uses `StringRef` objects (and not `ArrayRef<char>`), so yes, I think this should be a string.

cmtice commented on lldb/source/Core/DILAST.cpp:
```diff
@@ -0,0 +1,560 @@
+  return type.IsReferenceType() ? type.GetNonReferenceType() : type;
+}
+
+std::optional<MemberInfo>
```
I've tried that, but the lhs_val_sp does not always have the right type for finding the field name. :-(. One of the simplest examples: in one test case I have a class "C".  I declare an array variable "C c_arr[2];".  The ValueObject for 'c_arr' has the type "C[2]", rather than "C".  Type "C" has named field members, but type "C[2]" does not, so GetChildMemberWithName doesn't work on it.  I have *many* different test cases where it turns out the the ValueObject associated with the variable doesn't have quite the right type for using GetChildMemberWithName. By the time I've special-cased and updated the code to handle all of them, it doesn't look any better than the function I have here.

labath replied:
Why do we actually want this function to work on something like `C[2]`? That's definitely not what I would have expected from the function signature (nor from its implementation, actually).

Even that was the  behavior we want (which I really hope isn't the case), the other implementation might still make more sense, because it would give you a chance to explain what you're doing and why. Right now, this function looks like it's trying to reimplement `TypeSystemClang::GetIndexOfChildMemberWithName` without any indication of why. I wouldn't be surprised if it didn't work with other languages due to its reliance on how TypeSystemClang or C++ represent certain constructs.

(This extraction of information from the types could be part of why you're needing to detect if something is a smart pointer, etc. We currently don't have a way to tell if a type represents a smart pointer -- we can only do that for a specific value of a that type)

cmtice replied:
This function is called from the parser when it is trying to parse what ought to be a field/member name (i.e. it's gotten through <stuff-before-field-reference> <dot-or-arrow> field_name).  The "stuff-before-field-reference" can be a simple variable name, or something more complex; in this case it was an array subscript, 'c_arr[0]'. (The full command was 'frame variable c_arr[0].field_').  This function determines if the 'stuff-before-field-reference' has a valid field/member with the given name, and the index of that child/member (or set of potential valid indices) and returns that in the "idx" parameter, along with the full information about the field, which gets returned in the MemberInfo struct (the main return value).  This information is later used by the evaluator (once we have finished parsing the entire input) to actually find and return the value the user was asking for.

Since  whether or not the field/member name is valid really depends on the type, I do not understand why you seem to feel it is wrong to use the type (the second parameter) of the "stuff-before-field-reference" to check to see if it is valid and (if it's valid) what its position is?

cmtice committed:
[LLDB] Add AST node classes, functions, etc. for Data Inspection Language

Address latest round of review comments.

cmtice commented:
I think I have addressed all the review comments; please take another look. :-)

labath reviewed (commented):
Even though you've removed the smart pointer checking function (the most obvious example of language-specific behavior), I'm still worried about whether the overall design of this code will allow you to avoid it.

Looking at the `IdentifierInfo` class, I see that it can be constructed with only a type (the value is optional). I think this is going to be very hard to make work, since a type does not know if its a smart pointer (only a value does). Have you already figured out a way to make that work?

IndentifierInfo isn't the only suspicious place though. All of the places that are doing something with bitfield offsets are very suspicious to me, as that's something that the Values should already know.

I haven't seen the parser, but I wouldn't be surprised if this has something to do with the vexing parse issues of C languages, where one needs to know if something is a type just to parse the expression correctly. Now, if that's the case, and if really need to know if something is a smart pointer (i.e. whether it is dereferencable) without having access to a value of that type, then we may need to have a bigger discussion about what to do about that.

cmtice commented on lldb/include/lldb/Core/DILAST.h:
```diff
@@ -0,0 +1,459 @@
+#include "lldb/Symbol/TypeList.h"
+#include "lldb/Target/LanguageRuntime.h"
+#include "lldb/Utility/ConstString.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Basic/TokenKinds.h"
```
Doesn't clang compile all the languages we want to support?

labath replied:
Nope. Swift uses llvm, but not clang. The same goes for rust, and potentially other downstream languages.

labath replied:
Yeah, that does sound like a problem.

Putting this stuff in a plugin would be possible, but it would be a departure from what we've talked about earlier, since we would essentially end up with a different flavour of DIL for every language.

Saying that clang *is* the parser for the DIL might also be possible, but that would also warrant more discussion, because we have a (maybe undocumented?) rule of no clang dependencies on lldb core.

A third option might be to roll our own parser. Unlike lldb-eval, I think the DIL does not have the ambition of matching c++ (or any other language) precisely, so maybe we have some leeway to remove/skip/alter the features which make parsing c(++) the mess that it is.

werat replied:
DIL doesn't use Clang for parsing and have its own parser, that was the whole point of project :D
It does use Clang's lexer though, mostly for convenience, since it already has all the token types and provides source locations. If there's a strict rule of "not depending on clang libraries", then we can do it, but I'm not sure it's worth it.

labath replied:
Okay, so could one say that it's usage is similar to the code in [CPlusPlusNameParser](https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/Language/CPlusPlus/CPlusPlusNameParser.cpp), which uses clang to tokenize the (function) name, but then does the parsing itself? Or does it go deeper than that?

The thing I'm thinking about is that "having all of the tokens defined" may not be such an advantage if we later add some features from other languages (I don't know, let's say the `?.` operator from Swift). For lexing, that may not matter cause you could maybe lex `?` and `.` separately and then join them later, but if it's doing more, then that may be a problem..

cmtice replied:
Yes, Andy (werat) is correct. I mis-spoke in my previous comment (apologies!).  We use the Clang tokenizer & lexer but wrote our own parser.

Since that is the case, is this ok?

labath replied:
I think I'd be fine with that, but I also think it's not my call to make. @jimingham, @adrian-prantl, @bulbazord, what do you think?

bulbazord replied:
I haven't been following the evolution of this PR closely, but from this comment chain it looks like the question is "Can we have clang dependencies in lldbCore?". I won't gatekeep this PR for this reason, but here are my positions:
1. We should avoid having clang in lldbCore if possible.
2. I think you should write a lexer for the DIL.

As Pavel already pointed out, plenty of folks downstream want to use LLDB to debug their programs with languages other than C/C++/ObjC/ObjC++. Swift and Rust are great examples and on occasion I see people try to add minor changes to support their languages too. Adding clang as a core dependency makes it more difficult to decouple LLDB from C/C++ specifically. LLDB has the potential to be a more language-agnostic debugger where the core of the engine is flexible and focuses on operating on common abstractions while the Plugins dictate how to interact with the specific language/OS/platform. I think LLDB is still pretty far away from being that, but myself and other contributors have done a lot of work over the last few years to bring us closer to that reality.

As for the DIL, I think we should consider the goal. If the language is meant to use C/C++ syntax to inspect data, I suppose it makes sense to use clang's lexer to tokenize input. CPlusPlusNameParser uses clang because it's trying to parse C++ names. However, using clang now makes it harder to extend to support other languages in the future. What if clang doesn't chop up my input into tokens correctly or the way I want? Pavel mentioned a case where we can work around this, but the parser is just going to be workaround after workaround trying to fit a square peg into a round hole. You've already written the parser, what's a lexer too? :)

cmtice commented:
> Even though you've removed the smart pointer checking function (the most obvious example of language-specific behavior), I'm still worried about whether the overall design of this code will allow you to avoid it.
> 
I believe I have fixed all the code to avoid any need to to smart pointer checking -- I went back and looked very closely at what the current frame variable implementation does when handling smart pointers, and tried to do the same thing, which appears to have worked.

> Looking at the `IdentifierInfo` class, I see that it can be constructed with only a type (the value is optional). I think this is going to be very hard to make work, since a type does not know if its a smart pointer (only a value does). Have you already figured out a way to make that work?
> 
The case where it is constructed only with a type is meant to be when handling LLDB convenience variables (the dynamic variables names "$..."); I'm not sure if those can be smart pointers?

> IndentifierInfo isn't the only suspicious place though. All of the places that are doing something with bitfield offsets are very suspicious to me, as that's something that the Values should already know.
> 
> I haven't seen the parser, but I wouldn't be surprised if this has something to do with the vexing parse issues of C languages, where one needs to know if something is a type just to parse the expression correctly. Now, if that's the case, and if really need to know if something is a smart pointer (i.e. whether it is dereferencable) without having access to a value of that type, then we may need to have a bigger discussion about what to do about that.

I would be happy to share the parser code with you and walk/talk you through it, if that would help?

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-ast

cmtice committed:
[LLDB] Add AST node clases, functions, etc. for Data Inspection Language

Address review comments.

cmtice committed:
Fix clang format issues.

cmtice commented:
I think I have addressed all of the review comments now, except for the dependency on the Clang lexer.  If it's ok with the reviewers, I would rather not put that change into this PR, but instead I will make that the next PR that I submit in the DIL (Data Inspection Language) series.   However if you are strongly against that, I will add the new lexer to this PR as soon as I finish writing it.

So please review this again. Thank you!

cmtice commented:
Ping? Anybody?

Michael137 commented:
Is the plan to get the DIL data structures merged before the rest of the patch series is up for review? I think it'd be great to see how the infrastructure introduced here will be used

cmtice commented:
> Is the plan to get the DIL data structures merged before the rest of the patch series is up for review? I think it'd be great to see how the infrastructure introduced here will be used

There are 3 major pieces to the DIL implementation (soon to be 4, once I've got the lexer written):  The AST bit; the parser, which is a recursive descent parser using the .ebnf grammar; and the interpreter/evaluator.  The parser  performs parsing and type-checking on the input expression, building up the AST as it goes.  Once the AST is built, it gets passed to the interpreter/evaluator, which evaluates the AST (performing more correctness checks along the way), returning an LLDB ValueObjectSP at the end. I have both 'stripped down' and 'full' versions of all of these pieces. The 'stripped down' pieces match the current 'frame variable' functionality, and the 'full' versions implement extensions to that functionality (allowing it to handle all the expressions that lldb-eval could handle).

The plan is to submit a series of PRs as follows: 1). the AST parts (this PR).  2). the not-yet-written lexer (with any necessary changes to the already submitted AST parts); 3). the parser; 4) the evaluator;  5); the 'glue', i.e. updates to things like StackFrame.cpp & CommandObjectFrame.cpp, to  hook up the new DIL implementation (optionally), as well as the test cases.   Currently, the stripped down parser is ~2700 lines of code, and the stripped down evaluator is ~1000 lines of code.

As I mentioned in an earlier comment (on June 24), you can see the full (not stripped down) implementation at https://github.com/cmtice/llvm-project/tree/DIL-work-new/ , although I have not (yet) updated that with the updates that have gone into this PR.

Does that answer all of your questions? Or is there more you would like to know or see?

Michael137 commented:
> > Is the plan to get the DIL data structures merged before the rest of the patch series is up for review? I think it'd be great to see how the infrastructure introduced here will be used
> 
> There are 3 major pieces to the DIL implementation (soon to be 4, once I've got the lexer written): The AST bit; the parser, which is a recursive descent parser using the .ebnf grammar; and the interpreter/evaluator. The parser performs parsing and type-checking on the input expression, building up the AST as it goes. Once the AST is built, it gets passed to the interpreter/evaluator, which evaluates the AST (performing more correctness checks along the way), returning an LLDB ValueObjectSP at the end. I have both 'stripped down' and 'full' versions of all of these pieces. The 'stripped down' pieces match the current 'frame variable' functionality, and the 'full' versions implement extensions to that functionality (allowing it to handle all the expressions that lldb-eval could handle).
> 
> The plan is to submit a series of PRs as follows: 1). the AST parts (this PR). 2). the not-yet-written lexer (with any necessary changes to the already submitted AST parts); 3). the parser; 4) the evaluator; 5); the 'glue', i.e. updates to things like StackFrame.cpp & CommandObjectFrame.cpp, to hook up the new DIL implementation (optionally), as well as the test cases. Currently, the stripped down parser is ~2700 lines of code, and the stripped down evaluator is ~1000 lines of code.
> 
> As I mentioned in an earlier comment (on June 24), you can see the full (not stripped down) implementation at https://github.com/cmtice/llvm-project/tree/DIL-work-new/ , although I have not (yet) updated that with the updates that have gone into this PR.
> 
> Does that answer all of your questions? Or is there more you would like to know or see?

Thanks for the breakdown! I was just wondering whether the plan was to merge this before the other PRs are up. I think it'd be preferable to merge them once they all have been approved. Though if others are ok with the current strategy I don't want to block this.
